### PR TITLE
Feature rename uuid

### PR DESCRIPTION
**Pull request by rakshasa**
Renames a deleted item's uuid if a new item is created with the same uuid.

Includes nodeapi and endpoints refactoring and better error checking.

### CODE REVIEW

**Comment by rakshasa** [comment]
Fixes #490.

File: vnet/lib/vnet/endpoints/1.0/lease_policies.rb [hunk 1]
```diff
@@ -2,15 +2,16 @@
...
   def self.put_post_shared_params
-    param :mode, :String, in: CLP::MODES, default: CLP::MODE_SIMPLE
     param :timing, :String, in: CLP::TIMINGS, default: CLP::TIMING_IMMEDIATE
   end
 
   fill_options = [ ]
 
   put_post_shared_params
   param_uuid M::LeasePolicy
+  param :mode, :String, in: CLP::MODES, default: CLP::MODE_SIMPLE
```
Comment (Metallion) [inline vnet/lib/vnet/endpoints/1.0/lease_policies.rb]: This change needs to be made in vnctl too.

File: vnet/spec/vnet/endpoints/1.0/route_links_spec.rb [hunk 2]
```diff
@@ -32,7 +32,8 @@ def app
   end
 
   describe "PUT /:uuid" do
-    accepted_params = { :mac_address => "fe:17:9b:9f:e8:33" }
+    accepted_params = {
+    }
```
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/route_links_spec.rb]: And here too. There is no change to `endpoints/route_links` to match this. `:mac_address` is still an accepted parameter.

File: vnet/spec/vnet/endpoints/1.0/mac_lease_spec.rb [hunk 3]
```diff
@@ -41,7 +41,6 @@ def app
 
     accepted_params = {
       :interface_uuid => "if-test2",
-      :mac_address => "00:21:cc:da:e9:dc"
```
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/mac_lease_spec.rb]: `PUT mac_leases` still accepts a `:mac_address` parameter so it shouldn't be moved from the test.

File: vnet/spec/vnet/endpoints/1.0/network_services_spec.rb [hunk 4]
```diff
@@ -35,12 +35,10 @@ def app
 
   describe "PUT /:uuid" do
     let!(:new_interface) { Fabricate(:interface) { uuid "if-other"}  }
+
     accepted_params = {
-      :interface_uuid => "if-other",
-      :incoming_port => 40,
-      :outgoing_port => 100
```
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/network_services_spec.rb]: Here too these parameters are still accepted in the web api so they shouldn't be removed from the test.

File: vnet/spec/vnet/endpoints/1.0/networks_spec.rb [hunk 5]
```diff
@@ -38,10 +38,7 @@ def app
   describe "PUT /:uuid" do
     accepted_params = {
       :display_name => "our new name for the test network",
-      :ipv4_network => "10.0.0.2",
-      :ipv4_prefix => 8,
       :domain_name => "new.vdc.test.domain",
-      :network_mode => "physical",
```
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/networks_spec.rb]: Same again. As long as these parameters are still accepted by the webapi, we need to test them.

File: vnet/spec/vnet/endpoints/1.0/routes_spec.rb [hunk 6]
```diff
@@ -38,18 +38,9 @@ def app
...
-    let!(:new_interface) { Fabricate(:interface) { uuid 'if-newif' } }
-    let!(:new_network) { Fabricate(:network) { uuid "nw-newnw" } }
-    let!(:route_link) { Fabricate(:route_link) { uuid "rl-newroute" } }
-
     accepted_params = {
-      :interface_uuid => "if-newif",
-      :network_uuid => "nw-newnw",
-      :route_link_uuid => "rl-newroute",
-      :ipv4_network => "192.168.3.0",
-      :ipv4_prefix => 16,
```
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/routes_spec.rb]: And same thing again. The webapi still accepts these parameters so we need this test to confirm that they work.

**Comment by rakshasa** [comment]
Fixed all the issues.

File: vnet/spec/vnet/endpoints/1.0/routes_spec.rb (hunk 6 above)
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/routes_spec.rb]: Same comment as for `route_links`. Now that the PUT request no longer accepts any parameters, it doesn't do anything any more so I think we should remove it.

File: vnet/spec/vnet/endpoints/1.0/translations_spec.rb [hunk 7]
```diff
@@ -36,7 +36,6 @@ def app
     let!(:test_interface) { Fabricate(:interface, uuid: 'if-test2') }
 
     accepted_params = {
-      :interface_uuid => "if-test2"
```
Comment (Metallion) [inline vnet/spec/vnet/endpoints/1.0/translations_spec.rb]: I missed this one on my first review. This is also still in the webapi, as are `mode` and `passthrough`. They should also be tested or removed from the API.

**Comment by rakshasa** [comment]
Fixed.

**Pull request merged**